### ISSUE TITLE
Intl: Chakra does not correctly implement caseFirst support

### ISSUE THREAD

**Issue by jackhorton** [Microsoft/ChakraCore#4015]
In Edge 15:
```js
new Intl.Collator(undefined, { caseFirst: 'lower' }).compare("a", "A") // returns -1
new Intl.Collator(undefined, { caseFirst: 'upper' }).compare("a", "A") // returns -1
new Intl.Collator(undefined, { caseFirst: 'lower' }).resolvedOptions() // returns object with caseFirst: 'lower'
```

My understanding of the spec is that we are allowed to not support this option as long as any given values for caseFirst (or `-u-kf`) are validated _and_ we do not report caseFirst as a resolved option. @bterlson is this a correct interpretation? If so, we should have an intermediate fix of removing the caseFirst key from being reported by resolvedOptions. Regardless, we should look into implementing this in the Windows scenario.

**Comment by bterlson** [comment]
I'm not seeing anything in the spec that says what to do exactly if you don't support one of those options (in other words, we may be free to not support it and include it in resolved options). However, not including this in %collator%.[[RelevantExtensionKeys]] seems like the right choice, and that would have the effect of removing it from resolved options (but it would still be processed by collator initialization if passed as an extension or as an options property).

/cc @Caridy and @zbraniecki who are the super experts and get double-check me :)

**Comment by zbraniecki** [comment]
@littledan may know more, but my guess is that yes, you want to remove it from your `RelevantExtensionKeys`, and yes, you want to not report a resolvedOption for it.

The only nuance compared to what Brian said is that I think that those two things are not strongly connected. Not having a RelevantExtensionKey doesn't make it impossible to resolve `caseFirst` either by using the default or by respecting the `options` param.
As I said - that's a nuance - since you want to not support it completely, it should not be in resolvedOptions.

**Comment by bterlson** [comment]
@zbraniecki thanks so much for the info. To confirm, when you say "should not be in resolvedOptions", should I interpret "should" per RFC2119 or do you mean normatively/must? Just curious if I missed something in the spec. Even that should requirement (if it's an RFC2119 should) should be stated :)

**Comment by zbraniecki** [comment]
Definitely RFC2119 "should".

Quoting the spec:

```
A) Implementation Dependent Behaviour
Support for the Unicode extensions keys kn, kf and the parallel options properties numeric, caseFirst (10.1.1) 
```

```

10.3.5) Intl.Collator.prototype.resolvedOptions ()
(...)
The function returns a new object whose properties and attributes are set as (...) as well as those properties shown in Table 2 whose keys are included in the %Collator%.[[RelevantExtensionKeys]] internal slot of the standard built-in object that is the initial value of Intl.Collator.
```

So, my take is that the spec states that you should not return properties whos keys you do not include in RelevantExtensionKeys.

How would you want this to be expressed to make it less confusing?

**Comment by bterlson** [comment]
It's clear that if a key is not in RelevantExtensionKeys it doesn't show up on resolvedOptions. Where it gets muddled for me is the requirement that RelevantExtensionKeys should (or must?) not include keys for which the implementation doesn't use (10.2.3 says it may include kf, and 9.1 says that keys are "relevant for the functionality of constructed objects" which doesn't tell me much). Maybe I'm just missing something obvious :)

**Comment by littledan** [comment]
The spec is pretty explicit in the optionality of supporting caseFirst, e.g., [in the constructor](https://tc39.github.io/ecma402/#sec-intl.collator),

> 4. If %Collator%.[[RelevantExtensionKeys]] contains "kf", then
> a. Append [[CaseFirst]] as the last element of internalSlotsList.

Then, for the interpretation of CaseFirst, UTS 35 is normatively referenced from [here](https://tc39.github.io/ecma402/#sec-collator-compare-functions)

> For the interpretation of options settable through extension keys, see Unicode Technical Standard 35.

Although the comparison function is implementation-defined, it seems like only a somewhat twisted reading of all of this would lead you to say, "OK, the key is supported, with the meaning from UTS 35, and in particular, our implementation-defined semantics are to ignore it". I don't know whether these semantics would fall under "should not" or "must not", though.

For what it's worth, caseFirst isn't supported on all engines; on a version I tested, JSC and SpiderMonkey feature-test as unsupported through the easy one-liner `Intl.Collator().resolvedOptions().caseFirst === undefined`, and this is consistent with runtime behavior that ignores them. By contrast, V8 supports `caseFirst`, and it's also reflected in resolvedOptions. I like @jackhorton 's recommendation to remove it from resolvedOptions for now, as it will make feature testing work on all engines.

**Comment by dilijev** [comment]
@jackhorton is this fixed?

**Comment by jackhorton** [comment]
Closed in #4083
